<a href="https://colab.research.google.com/github/prashanthpasikanti/Mobile-Price-Prediction/blob/main/DS_Minor_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df=pd.read_csv("/content/mobile_price_range_data.csv")

In [ ]:
df


battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0               842     0          2.2         0   1       0           7   
1              1021     1          0.5         1   0       1          53   
2               563     1          0.5         1   2       1          41   
3               615     1          2.5         0   0       0          10   
4              1821     1          1.2         0  13       1          44   
...             ...   ...          ...       ...  ..     ...         ...   
1995            794     1          0.5         1   0       1           2   
1996           1965     1          2.6         1   0       0          39   
1997           1911     0          0.9         1   1       1          36   
1998           1512     0          0.9         0   4       1          46   
1999            510     1          2.0         1   5       1          45   

      m_dep  mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  \
0       0.6        188        2  ...         20       756  2549     9     7   
1       0.7        136        3  ...        905      1988  2631    17     3   
2       0.9        145        5  ...       1263      1716  2603    11     2   
3       0.8        131        6  ...       1216      1786  2769    16     8   
4       0.6        141        2  ...       1208      1212  1411     8     2   
...     ...        ...      ...  ...        ...       ...   ...   ...   ...   
1995    0.8        106        6  ...       1222      1890   668    13     4   
1996    0.2        187        4  ...        915      1965  2032    11    10   
1997    0.7        108        8  ...        868      1632  3057     9     1   
1998    0.1        145        5  ...        336       670   869    18    10   
1999    0.9        168        6  ...        483       754  3919    19     4   

      talk_time  three_g  touch_screen  wifi  price_range  
0            19        0             0     1            1  
1             7        1             1     0            2  
2             9        1             1     0            2  
3            11        1             0     0            2  
4            15        1             1     0            1  
...         ...      ...           ...   ...          ...  
1995         19        1             1     0            0  
1996         16        1             1     1            2  
1997          5        1             1     0            3  
1998         19        1             1     1            0  
1999          2        1             1     1            3  

[2000 rows x 21 columns]

In [ ]:
df.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

In [ ]:
df.shape

(2000, 21)

In [ ]:
df.isnull().sum()

battery_power    0
blue             0
clock_speed      0
dual_sim         0
fc               0
four_g           0
int_memory       0
m_dep            0
mobile_wt        0
n_cores          0
pc               0
px_height        0
px_width         0
ram              0
sc_h             0
sc_w             0
talk_time        0
three_g          0
touch_screen     0
wifi             0
price_range      0
dtype: int64

In [ ]:
x=df[['battery_power','blue','clock_speed','dual_sim','fc','four_g','int_memory','m_dep','mobile_wt','n_cores','pc','px_height','px_width','ram','sc_h','sc_w','talk_time','three_g','touch_screen','wifi']]
y=df['price_range']
print(type(x))
print(type(y))


<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>


In [ ]:
from sklearn.model_selection import train_test_split


In [ ]:
print(x.shape)
print(y.shape)

(2000, 20)
(2000,)


In [ ]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25)

In [ ]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(1500, 20)
(500, 20)
(1500,)
(500,)


In [ ]:
x_train.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
1223            514     1          1.6         0   7       1          37   
885             730     1          1.3         1   6       0           6   
867            1498     1          0.7         0   3       1           8   
1976           1201     1          0.5         0   2       0          10   
1449            602     0          0.9         0   9       0          50   

      m_dep  mobile_wt  n_cores  pc  px_height  px_width   ram  sc_h  sc_w  \
1223    0.1        172        1   9        956      1723  3392    12     8   
885     0.9        167        7  10       1096      1155  2812    10     1   
867     0.1        170        7   4        347      1076  3358     7     3   
1976    1.0         99        7   7        306       558   495    15     6   
1449    0.9        141        8  18       1571      1717  2189     7     5   

      talk_time  three_g  touch_screen  wifi  
1223          5        1             1     1  
885          17        0             0     0  
867          19        1             0     0  
1976         14        1             1     1  
1449         16        0             1     0

**a)Logistic Regression**

In [ ]:
from sklearn.linear_model import LogisticRegression


In [ ]:
m1=LogisticRegression()
m1.fit(x_train,y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


LogisticRegression()

In [ ]:
ypred_m1=m1.predict(x_test)
print(ypred_m1)

[2 3 1 1 0 1 3 2 2 3 3 2 1 1 0 1 2 2 0 0 3 1 0 2 3 3 3 0 3 0 2 3 2 3 1 0 3
 3 0 3 3 1 2 2 3 2 3 2 3 1 1 1 1 0 2 1 3 1 3 1 3 3 1 3 2 1 0 0 3 0 1 0 0 3
 0 0 3 0 1 0 1 0 2 0 1 0 1 1 2 1 2 1 3 2 2 1 3 0 2 1 2 2 0 2 2 2 2 2 3 2 0
 1 1 0 1 2 0 1 2 2 1 0 3 2 2 0 3 2 2 1 3 2 1 3 3 2 1 3 1 2 1 2 1 2 1 1 1 1
 1 3 3 0 0 1 3 0 2 3 1 0 3 1 2 0 3 1 2 0 2 2 1 1 1 0 2 2 2 3 2 1 3 1 0 0 0
 1 1 3 2 3 0 1 2 0 1 0 2 2 1 2 3 3 2 0 1 3 0 3 0 2 2 0 2 3 1 2 3 3 2 1 3 2
 3 3 0 3 0 3 1 3 0 0 0 1 1 2 1 0 0 0 0 1 1 0 1 2 3 2 3 2 0 3 2 0 1 2 0 0 2
 2 3 3 0 3 0 0 3 3 3 1 3 3 0 1 2 0 2 0 3 1 2 3 0 3 0 1 3 0 3 3 2 3 3 2 3 0
 3 2 3 0 2 3 3 3 2 2 0 3 1 0 2 0 1 2 3 0 3 0 2 1 0 0 0 1 0 0 2 1 0 2 3 3 0
 0 0 0 3 3 3 3 2 0 2 1 3 2 1 2 1 3 2 3 1 1 0 2 3 2 1 3 2 3 2 1 2 2 1 1 3 3
 0 2 1 2 0 0 1 1 3 3 2 0 0 0 2 2 2 0 0 0 1 2 2 1 0 2 3 0 1 3 3 0 3 1 1 2 1
 2 1 0 3 0 0 0 0 0 2 0 3 3 0 1 2 0 1 3 2 0 3 0 1 3 3 0 2 1 3 1 3 3 3 3 3 2
 1 3 1 2 3 0 3 3 2 2 1 2 2 1 2 1 3 0 0 2 1 1 2 1 3 3 1 1 2 3 0 1 0 2 2 3 1
 3 3 0 2 3 0 2 3 2 3 0 2 

In [ ]:
from sklearn.metrics import confusion_matrix,classification_report

In [ ]:
cm1=confusion_matrix(y_test,ypred_m1)
print(cm1)
print(classification_report(y_test,ypred_m1))

[[103  17   1   0]
 [ 18  76  43   6]
 [  1  23  56  36]
 [  0   1  27  92]]
              precision    recall  f1-score   support

           0       0.84      0.85      0.85       121
           1       0.65      0.53      0.58       143
           2       0.44      0.48      0.46       116
           3       0.69      0.77      0.72       120

    accuracy                           0.65       500
   macro avg       0.66      0.66      0.65       500
weighted avg       0.66      0.65      0.65       500



Accuracy for Logistic Regression is 0.61

**b)KNN Classification**

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
m2=KNeighborsClassifier(n_neighbors=20)
m2.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=20)

In [ ]:
ypred_m2 = m2.predict(x_test)
print(ypred_m2)

[1 3 1 1 1 1 3 2 3 2 3 3 0 1 0 0 3 2 0 0 3 1 0 2 3 2 2 0 3 0 1 3 2 3 0 0 3
 2 0 3 3 2 3 2 2 1 3 1 3 1 0 1 1 0 1 1 2 1 3 1 3 3 1 3 1 2 0 0 2 0 1 0 0 3
 0 0 3 0 1 0 0 0 3 0 2 0 1 2 2 1 1 1 2 1 2 1 3 1 3 2 2 1 0 1 3 3 2 2 3 2 0
 2 2 0 1 3 0 0 2 2 1 0 3 2 2 1 3 1 1 2 2 1 1 3 3 0 1 3 0 2 1 1 1 2 1 1 1 1
 1 3 2 0 0 1 3 0 2 3 0 0 2 2 2 0 3 1 3 0 2 1 1 0 0 0 2 2 3 3 1 1 2 1 0 0 0
 1 1 2 1 3 0 1 2 0 2 0 2 2 0 3 3 1 2 0 1 2 0 2 0 1 2 0 2 3 1 1 3 3 3 1 1 2
 3 3 0 3 0 2 0 2 0 1 0 1 1 1 1 0 0 0 1 1 0 0 1 3 3 1 3 3 0 3 2 0 0 2 0 0 2
 1 2 3 0 3 0 0 3 3 2 1 2 3 0 1 1 0 1 1 2 0 3 3 0 2 1 2 2 0 2 3 3 3 3 2 3 0
 3 2 2 0 2 3 3 3 3 2 0 2 2 0 2 0 1 1 2 0 2 0 2 1 1 0 0 1 0 0 1 1 0 1 3 3 0
 1 1 0 3 2 3 1 2 0 1 2 3 3 2 2 1 1 2 3 1 1 0 1 2 1 1 3 1 3 1 2 3 3 0 0 2 3
 0 2 1 2 0 0 2 3 2 3 1 0 0 0 2 2 1 1 0 0 1 2 2 0 1 1 3 0 1 3 3 0 3 1 1 1 2
 1 1 1 2 0 1 0 0 1 1 0 2 3 0 1 2 0 1 3 2 0 3 1 1 3 2 0 2 1 3 1 2 3 3 3 3 1
 1 2 2 2 3 0 3 3 3 3 1 2 2 0 3 1 3 0 0 1 1 0 2 1 3 2 0 1 2 3 1 1 0 2 2 3 1
 2 3 0 3 3 1 2 3 2 1 0 3 

In [ ]:
cm2=confusion_matrix(y_test,ypred_m2)
print(cm2)

[[120   1   0   0]
 [  6 131   6   0]
 [  0   9 103   4]
 [  0   0   8 112]]


In [ ]:
print(classification_report(y_test,ypred_m2))

              precision    recall  f1-score   support

           0       0.95      0.99      0.97       121
           1       0.93      0.92      0.92       143
           2       0.88      0.89      0.88       116
           3       0.97      0.93      0.95       120

    accuracy                           0.93       500
   macro avg       0.93      0.93      0.93       500
weighted avg       0.93      0.93      0.93       500



Accuracy for KNN Classification model is 0.93

**c)SVM with Linear kernel**

In [ ]:
from sklearn.svm import SVC

In [ ]:
m3=SVC(kernel='linear',C=1)
m3.fit(x_train,y_train)

SVC(C=1, kernel='linear')

In [ ]:
ypred_m3=m3.predict(x_test)
print(ypred_m3)

[2 3 1 1 1 1 3 2 3 2 3 2 0 1 0 0 3 2 0 0 3 1 0 2 3 3 2 0 3 0 2 3 1 3 0 0 3
 2 0 3 3 2 3 2 2 1 3 1 3 1 0 1 2 0 1 1 2 1 3 1 3 3 1 3 1 2 0 0 2 0 1 0 0 3
 0 0 2 0 1 0 0 0 3 0 1 0 1 2 2 1 2 1 2 1 2 1 3 1 2 2 2 1 0 1 3 3 2 2 3 2 0
 2 2 0 2 3 0 0 2 2 1 1 3 2 2 1 3 1 1 2 3 1 1 3 3 1 1 3 0 2 1 1 1 3 1 1 1 1
 1 3 3 0 0 1 3 0 2 3 0 0 2 2 2 0 3 1 3 0 2 1 1 0 0 0 1 2 3 3 1 1 2 1 0 0 0
 1 1 2 1 3 0 1 2 0 2 0 2 2 0 3 3 1 1 0 2 2 0 2 0 1 2 0 2 3 1 1 3 3 3 1 1 2
 3 3 0 3 0 3 0 2 0 1 0 1 1 1 1 0 0 0 1 1 0 0 1 3 3 1 3 3 0 3 1 1 0 2 0 0 2
 1 2 3 0 3 0 0 3 3 2 1 2 3 0 1 1 0 1 1 2 0 3 3 0 2 1 2 2 0 2 3 3 3 3 2 3 0
 3 2 2 0 2 3 3 3 3 2 0 2 2 0 2 0 1 1 2 0 2 0 2 1 1 0 0 1 0 0 1 1 0 1 3 3 0
 1 1 0 3 2 3 1 2 0 1 2 3 3 2 2 1 1 2 3 1 1 0 1 3 1 1 3 1 3 1 2 3 3 0 0 2 3
 0 1 2 2 0 0 1 3 2 3 1 0 0 0 2 1 1 1 0 0 1 2 2 0 1 1 3 0 1 3 3 0 3 1 1 1 2
 2 1 1 2 0 1 0 0 1 1 0 2 3 0 1 2 0 1 3 2 0 3 1 1 3 2 0 2 1 3 1 1 3 3 3 3 1
 1 2 2 2 3 0 3 3 3 3 1 2 2 0 3 1 3 0 0 1 1 0 2 1 3 3 0 1 2 3 1 1 0 2 2 3 1
 2 3 0 3 3 1 2 3 2 1 0 3 

In [ ]:
cm3=confusion_matrix(y_test,ypred_m3)
print(cm3)
print(classification_report(y_test,ypred_m3))

[[119   2   0   0]
 [  4 137   2   0]
 [  0   5 107   4]
 [  0   0   4 116]]
              precision    recall  f1-score   support

           0       0.97      0.98      0.98       121
           1       0.95      0.96      0.95       143
           2       0.95      0.92      0.93       116
           3       0.97      0.97      0.97       120

    accuracy                           0.96       500
   macro avg       0.96      0.96      0.96       500
weighted avg       0.96      0.96      0.96       500



Accuracy for SVM(Linear) is 0.96

**d)SVM with rbf kernel**

In [ ]:
m4=SVC(kernel='rbf',gamma=0.1,C=0.1)
m4.fit(x_train,y_train)

SVC(C=0.1, gamma=0.1)

In [ ]:
ypred_m4=m4.predict(x_test)
print(ypred_m4)

[2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 

In [ ]:
cm4=confusion_matrix(y_test,ypred_m4)
print(cm4)
print(classification_report(y_test,ypred_m4))

[[  0   0 121   0]
 [  0   0 143   0]
 [  0   0 116   0]
 [  0   0 120   0]]
              precision    recall  f1-score   support

           0       0.00      0.00      0.00       121
           1       0.00      0.00      0.00       143
           2       0.23      1.00      0.38       116
           3       0.00      0.00      0.00       120

    accuracy                           0.23       500
   macro avg       0.06      0.25      0.09       500
weighted avg       0.05      0.23      0.09       500



/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Accuracy of SVM(rbf kernel) is 0.23

**So, The best selected model is SVM(linear) with highest accuracy of 0.96**